// webpack-visualizer-plugin2/lib/plugin.js
"use strict";

var _path = _interopRequireDefault(require("path"));
var _fs = _interopRequireDefault(require("fs"));
var _mkdirp = _interopRequireDefault(require("mkdirp"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { "default": e }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
var cssString = _fs["default"].readFileSync(_path["default"].join(__dirname, './style.css'), 'utf8');
var jsString = _fs["default"].readFileSync(_path["default"].join(__dirname, './main.js'), 'utf8');
module.exports = /*#__PURE__*/function () {
  function VisualizerPlugin() {
    var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var statOpts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
    _classCallCheck(this, VisualizerPlugin);
    this.opts = _objectSpread({
      filename: 'stats.html',
      throwOnError: true
    }, opts);
    this.statOpts = _objectSpread({
      chunkModules: true
    }, statOpts);
  }
  return _createClass(VisualizerPlugin, [{
    key: "apply",
    value: function apply(compiler) {
      var _this = this;
      compiler.hooks.emit.tapAsync('Visualizer', function (compilation, callback) {
        var html;
        try {
          var stats = compilation.getStats().toJson(_this.statOpts);
          var stringifiedStats = JSON.stringify(stats).replace(/</g, '&lt;').replace(/>/g, '&gt;');
          html = "\n                    <!DOCTYPE html>\n                    <html lang=\"en\">\n                        <head>\n                            <meta charset=\"UTF-8\">\n                            <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n                            <title>Webpack Visualizer</title>\n                            <style>".concat(cssString, "</style>\n                        </head>\n                        <body>\n                            <div id=\"App\"></div>\n                            <script>window.stats = ").concat(stringifiedStats, ";</script>\n                            <script>").concat(jsString, "</script>\n                        </body>\n                    </html>\n                ");
        } catch (error) {
          console.error('webpack-visualizer-plugin: error creating stats file');
          if (_this.opts.throwOnError) {
            return callback(error);
          } else {
            console.error(error);
            return callback();
          }
        }
        var outputFile = _path["default"].join(compilation.outputOptions.path, _this.opts.filename);
        (0, _mkdirp["default"])(_path["default"].dirname(outputFile), function (mkdirpErr) {
          if (mkdirpErr) {
            console.error('webpack-visualizer-plugin: error writing stats file');
            if (_this.opts.throwOnError) {
              return callback(mkdirpErr);
            } else {
              return callback();
            }
          }
          _fs["default"].writeFile(outputFile, html, function (err) {
            if (err) {
              console.error('webpack-visualizer-plugin: error writing stats file');
              if (_this.opts.throwOnError) {
                return callback(err);
              }
            }
            callback();
          });
        });
      });
    }
  }]);
}();
